autoemulate/hyperparam_searching.py

```diff
@@ -0,0 +1,175 @@
+import logging
+from sklearn.model_selection import RandomizedSearchCV
+from skopt import BayesSearchCV
+
+from autoemulate.utils import (
+    get_model_name,
+    get_model_params,
+    get_model_param_grid,
+    adjust_param_grid,
+)
+
+
+def optimize_params(
+    X,
+    y,
+    cv,
+    model,
+    search_type="random",
+    niter=20,
+    param_grid=None,
+    n_jobs=None,
+    logger=None,
+):
+    """Performs hyperparameter search for the provided model.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Simulation input samples.
+    y : array-like, shape (n_samples,) or (n_samples, n_outputs)
+        Simulation output.
+    cv : int, cross-validation generator or an iterable
+        Determines the cross-validation splitting strategy.
+    model : model instance to do hyperparameter search for.
+    search_type : str, default="random"
+        Type of search to perform. Can be "random" or "bayes", "grid" not yet implemented.
+    niter : int, default=20
+        Number of parameter settings that are sampled. Trades off runtime vs quality of the solution.
+        param_grid : dict, default=None
+        Dictionary with parameters names (string) as keys and lists of
+        parameter settings to try as values, or a list of such dictionaries,
+        in which case the grids spanned by each dictionary in the list are
+        explored. This enables searching over any sequence of parameter
+        settings. Parameters names should be prefixed with "model__" to indicate that
+        they are parameters of the model.
+    n_jobs : int
+        Number of jobs to run in parallel.
+    logger : logging.Logger
+        Logger instance.
+
+    Returns
+    -------
+    searcher : sklearn.model_selection._search.BaseSearchCV or skopt.searchcv.BayesSearchCV
+        Searcher instance.
+    """
+    model_name = get_model_name(model)
+    logger.info(f"Performing grid search for {model_name}...")
+    param_grid = process_param_grid(model, search_type, param_grid)
+
+    # random search
+    if search_type == "random":
+        searcher = RandomizedSearchCV(
+            model,
+            param_grid,
+            n_iter=niter,
+            cv=cv,
+            n_jobs=n_jobs,
+            refit=True,
+        )
+    # Bayes search
+    elif search_type == "bayes":
+        searcher = BayesSearchCV(
+            model,
+            param_grid,
+            n_iter=niter,
+            cv=cv,
+            n_jobs=n_jobs,
+            refit=True,
+        )
+    elif search_type == "grid":
+        raise NotImplementedError("Grid search not available yet.")
```

Comment:
Probably want an ``else`` here to raise a more useful error message if ``search_type`` isn't an allowed value. You may also want to something like converting the string to lower case to be a bit more tolerant of the range of inputs you can accept. You may also want to permit passing a sklearn CV instance directly if they want to configure it themselves.

Reply:
Thanks, done. Allowing to pass a sklearn cv instance is already issue #116 .